// @umijs/core/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnableBy = exports.ApplyPluginsType = exports.ConfigChangeType = exports.ServiceStage = exports.PluginType = exports.Env = void 0;
var Env;
(function (Env) {
    Env["development"] = "development";
    Env["production"] = "production";
    Env["test"] = "test";
})(Env = exports.Env || (exports.Env = {}));
var PluginType;
(function (PluginType) {
    PluginType["preset"] = "preset";
    PluginType["plugin"] = "plugin";
})(PluginType = exports.PluginType || (exports.PluginType = {}));
var ServiceStage;
(function (ServiceStage) {
    ServiceStage[ServiceStage["uninitialized"] = 0] = "uninitialized";
    ServiceStage[ServiceStage["init"] = 1] = "init";
    ServiceStage[ServiceStage["initPresets"] = 2] = "initPresets";
    ServiceStage[ServiceStage["initPlugins"] = 3] = "initPlugins";
    ServiceStage[ServiceStage["resolveConfig"] = 4] = "resolveConfig";
    ServiceStage[ServiceStage["collectAppData"] = 5] = "collectAppData";
    ServiceStage[ServiceStage["onCheck"] = 6] = "onCheck";
    ServiceStage[ServiceStage["onStart"] = 7] = "onStart";
    ServiceStage[ServiceStage["runCommand"] = 8] = "runCommand";
})(ServiceStage = exports.ServiceStage || (exports.ServiceStage = {}));
var ConfigChangeType;
(function (ConfigChangeType) {
    ConfigChangeType["reload"] = "reload";
    ConfigChangeType["regenerateTmpFiles"] = "regenerateTmpFiles";
})(ConfigChangeType = exports.ConfigChangeType || (exports.ConfigChangeType = {}));
var ApplyPluginsType;
(function (ApplyPluginsType) {
    ApplyPluginsType["add"] = "add";
    ApplyPluginsType["modify"] = "modify";
    ApplyPluginsType["event"] = "event";
})(ApplyPluginsType = exports.ApplyPluginsType || (exports.ApplyPluginsType = {}));
var EnableBy;
(function (EnableBy) {
    EnableBy["register"] = "register";
    EnableBy["config"] = "config";
})(EnableBy = exports.EnableBy || (exports.EnableBy = {}));
