// @umijs/utils/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.yParser = exports.updatePackageJSON = exports.stripAnsi = exports.semver = exports.rimraf = exports.resolve = exports.prompts = exports.portfinder = exports.pkgUp = exports.Mustache = exports.logger = exports.lodash = exports.installDeps = exports.generateFile = exports.BaseGenerator = exports.Generator = exports.glob = exports.fsExtra = exports.deepmerge = exports.debug = exports.crossSpawn = exports.chokidar = exports.cheerio = exports.chalk = exports.axios = exports.address = exports.register = void 0;
const chokidar = __importStar(require("chokidar"));
exports.chokidar = chokidar;
const address_1 = __importDefault(require("../compiled/address"));
exports.address = address_1.default;
const axios_1 = __importDefault(require("../compiled/axios"));
exports.axios = axios_1.default;
const chalk_1 = __importDefault(require("../compiled/chalk"));
exports.chalk = chalk_1.default;
const cheerio_1 = __importDefault(require("../compiled/cheerio"));
exports.cheerio = cheerio_1.default;
const cross_spawn_1 = __importDefault(require("../compiled/cross-spawn"));
exports.crossSpawn = cross_spawn_1.default;
const debug_1 = __importDefault(require("../compiled/debug"));
exports.debug = debug_1.default;
const deepmerge_1 = __importDefault(require("../compiled/deepmerge"));
exports.deepmerge = deepmerge_1.default;
const fs_extra_1 = __importDefault(require("../compiled/fs-extra"));
exports.fsExtra = fs_extra_1.default;
const glob_1 = __importDefault(require("../compiled/glob"));
exports.glob = glob_1.default;
// import globby from '../compiled/globby';
const lodash_1 = __importDefault(require("../compiled/lodash"));
exports.lodash = lodash_1.default;
const mustache_1 = __importDefault(require("../compiled/mustache"));
exports.Mustache = mustache_1.default;
const pkgUp = __importStar(require("../compiled/pkg-up"));
exports.pkgUp = pkgUp;
const portfinder_1 = __importDefault(require("../compiled/portfinder"));
exports.portfinder = portfinder_1.default;
const prompts_1 = __importDefault(require("../compiled/prompts"));
exports.prompts = prompts_1.default;
const resolve_1 = __importDefault(require("../compiled/resolve"));
exports.resolve = resolve_1.default;
const rimraf_1 = __importDefault(require("../compiled/rimraf"));
exports.rimraf = rimraf_1.default;
const semver_1 = __importDefault(require("../compiled/semver"));
exports.semver = semver_1.default;
const strip_ansi_1 = __importDefault(require("../compiled/strip-ansi"));
exports.stripAnsi = strip_ansi_1.default;
const yargs_parser_1 = __importDefault(require("../compiled/yargs-parser"));
exports.yParser = yargs_parser_1.default;
const BaseGenerator_1 = __importDefault(require("./BaseGenerator/BaseGenerator"));
exports.BaseGenerator = BaseGenerator_1.default;
const generateFile_1 = __importDefault(require("./BaseGenerator/generateFile"));
exports.generateFile = generateFile_1.default;
const Generator_1 = __importDefault(require("./Generator/Generator"));
exports.Generator = Generator_1.default;
const installDeps_1 = __importDefault(require("./installDeps"));
exports.installDeps = installDeps_1.default;
const logger = __importStar(require("./logger"));
exports.logger = logger;
const updatePackageJSON_1 = __importDefault(require("./updatePackageJSON"));
exports.updatePackageJSON = updatePackageJSON_1.default;
__exportStar(require("./getCorejsVersion"), exports);
__exportStar(require("./importLazy"), exports);
__exportStar(require("./isStyleFile"), exports);
__exportStar(require("./npmClient"), exports);
__exportStar(require("./randomColor/randomColor"), exports);
exports.register = __importStar(require("./register"));
__exportStar(require("./tryPaths"), exports);
__exportStar(require("./winPath"), exports);
