// @umijs/utils/dist/npmClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installWithNpmClient = exports.checkNpmClient = exports.getNpmClient = exports.NpmClientEnum = exports.npmClients = void 0;
exports.npmClients = ['pnpm', 'tnpm', 'cnpm', 'yarn', 'npm'];
var NpmClientEnum;
(function (NpmClientEnum) {
    NpmClientEnum["pnpm"] = "pnpm";
    NpmClientEnum["tnpm"] = "tnpm";
    NpmClientEnum["cnpm"] = "cnpm";
    NpmClientEnum["yarn"] = "yarn";
    NpmClientEnum["npm"] = "npm";
})(NpmClientEnum = exports.NpmClientEnum || (exports.NpmClientEnum = {}));
const getNpmClient = () => {
    const userAgent = process.env.npm_config_user_agent;
    if (userAgent) {
        if (userAgent.includes('pnpm'))
            return 'pnpm';
        if (userAgent.includes('tnpm'))
            return 'tnpm';
        if (userAgent.includes('yarn'))
            return 'yarn';
        if (userAgent.includes('cnpm'))
            return 'cnpm';
    }
    return 'npm';
};
exports.getNpmClient = getNpmClient;
const checkNpmClient = (npmClient) => {
    const userAgent = process.env.npm_config_user_agent;
    return !!(userAgent && userAgent.includes(npmClient));
};
exports.checkNpmClient = checkNpmClient;
const installWithNpmClient = ({ npmClient, cwd, }) => {
    const { sync } = require('../compiled/cross-spawn');
    const npm = sync(npmClient, [npmClient === 'yarn' ? '' : 'install'], {
        stdio: 'inherit',
        cwd,
    });
    if (npm.error) {
        throw npm.error;
    }
};
exports.installWithNpmClient = installWithNpmClient;
