// @umijs/test/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConfig = void 0;
function getJSTransformer(jsTransformer) {
    switch (jsTransformer) {
        case 'esbuild':
            return require.resolve('esbuild-jest');
        case 'swc':
            return require.resolve('@swc-node/jest');
        case 'ts-jest':
            return require.resolve('ts-jest');
        default:
            throw new Error(`Unknown jsTransformer: ${jsTransformer}`);
    }
}
function createConfig(opts) {
    const config = {
        testMatch: ['**/*.test.(t|j)s(x)?'],
        transform: {
            '^.+\\.tsx?$': getJSTransformer((opts === null || opts === void 0 ? void 0 : opts.jsTransformer) || 'esbuild'),
        },
        moduleNameMapper: {
            '^.+\\.(css|less|sass|scss|stylus)$': require.resolve('identity-obj-proxy'),
        },
        testTimeout: 30000,
        modulePathIgnorePatterns: [
            '<rootDir>/packages/.+/compiled',
            '<rootDir>/packages/.+/fixtures',
        ],
        setupFiles: [require.resolve('../setupFiles/shim')],
    };
    if ((opts === null || opts === void 0 ? void 0 : opts.target) === 'browser') {
        config.testEnvironment = 'jsdom';
    }
    return config;
}
exports.createConfig = createConfig;
