// @umijs/ast/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generate = exports.setConfigByName = exports.removeConfigByName = exports.isReactComponent = exports.getExportProps = exports.getASTByFilePath = void 0;
var getASTByFilePath_1 = require("./getASTByFilePath/getASTByFilePath");
Object.defineProperty(exports, "getASTByFilePath", { enumerable: true, get: function () { return getASTByFilePath_1.getASTByFilePath; } });
var getExportProps_1 = require("./getExportProps/getExportProps");
Object.defineProperty(exports, "getExportProps", { enumerable: true, get: function () { return getExportProps_1.getExportProps; } });
var isReactComponent_1 = require("./isReactComponent/isReactComponent");
Object.defineProperty(exports, "isReactComponent", { enumerable: true, get: function () { return isReactComponent_1.isReactComponent; } });
var removeConfigByName_1 = require("./removeConfigByName/removeConfigByName");
Object.defineProperty(exports, "removeConfigByName", { enumerable: true, get: function () { return removeConfigByName_1.removeConfigByName; } });
var setConfigByName_1 = require("./setConfigByName/setConfigByName");
Object.defineProperty(exports, "setConfigByName", { enumerable: true, get: function () { return setConfigByName_1.setConfigByName; } });
var generate_1 = require("./utils/generate");
Object.defineProperty(exports, "generate", { enumerable: true, get: function () { return generate_1.generate; } });
