// @umijs/preset-umi/dist/features/configPlugins/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSchemas = void 0;
const utils_1 = require("@umijs/utils");
function getSchemas() {
    return {
        base: (Joi) => Joi.string(),
        conventionRoutes: (Joi) => Joi.object({
            base: Joi.string(),
            exclude: Joi.array().items(Joi.any()),
        }),
        headScripts: (Joi) => Joi.array(),
        history: (Joi) => Joi.object({
            type: Joi.string().valid('browser', 'hash', 'memory'),
        }),
        links: (Joi) => Joi.array(),
        metas: (Joi) => Joi.array(),
        mountElementId: (Joi) => Joi.string(),
        npmClient: (Joi) => Joi.string().valid(utils_1.NpmClientEnum.pnpm, utils_1.NpmClientEnum.tnpm, utils_1.NpmClientEnum.cnpm, utils_1.NpmClientEnum.yarn, utils_1.NpmClientEnum.npm),
        plugins: (Joi) => Joi.array().items(Joi.string()),
        publicPath: (Joi) => Joi.string().regex(/\/$/).error(new Error('publicPath must end with /')),
        routes: (Joi) => Joi.array().items(Joi.object()),
        scripts: (Joi) => Joi.array(),
        styles: (Joi) => Joi.array(),
    };
}
exports.getSchemas = getSchemas;
