// @umijs/bundler-webpack/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CSSMinifier = exports.JSMinifier = exports.Transpiler = exports.Env = void 0;
var Env;
(function (Env) {
    Env["development"] = "development";
    Env["production"] = "production";
})(Env = exports.Env || (exports.Env = {}));
var Transpiler;
(function (Transpiler) {
    Transpiler["babel"] = "babel";
    Transpiler["swc"] = "swc";
    Transpiler["esbuild"] = "esbuild";
    Transpiler["none"] = "none";
})(Transpiler = exports.Transpiler || (exports.Transpiler = {}));
var JSMinifier;
(function (JSMinifier) {
    JSMinifier["terser"] = "terser";
    JSMinifier["swc"] = "swc";
    JSMinifier["esbuild"] = "esbuild";
    JSMinifier["uglifyJs"] = "uglifyJs";
    JSMinifier["none"] = "none";
})(JSMinifier = exports.JSMinifier || (exports.JSMinifier = {}));
var CSSMinifier;
(function (CSSMinifier) {
    CSSMinifier["esbuild"] = "esbuild";
    CSSMinifier["cssnano"] = "cssnano";
    CSSMinifier["parcelCSS"] = "parcelCSS";
    CSSMinifier["none"] = "none";
})(CSSMinifier = exports.CSSMinifier || (exports.CSSMinifier = {}));
