// @umijs/bundler-esbuild/dist/utils/postcssProcess.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.postcssProcess = void 0;
const postcss_1 = __importDefault(require("postcss"));
const getBrowserlist_1 = require("./getBrowserlist");
function postcssProcess(config, css, path) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield (0, postcss_1.default)([
            require('postcss-flexbugs-fixes'),
            require('postcss-preset-env')({
                browsers: (0, getBrowserlist_1.getBrowserlist)((config === null || config === void 0 ? void 0 : config.targets) || {}),
                autoprefixer: Object.assign({ flexbox: 'no-2009' }, config === null || config === void 0 ? void 0 : config.autoprefixer),
                stage: 3,
            }),
        ].concat((config === null || config === void 0 ? void 0 : config.extraPostCSSPlugins) || [])).process(css, {
            from: path,
            to: path,
        });
    });
}
exports.postcssProcess = postcssProcess;
