// @umijs/bundler-webpack/dist/config/bundleAnalyzerPlugin.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addBundleAnalyzerPlugin = void 0;
// @ts-ignore
const webpack_bundle_analyzer_1 = require("@umijs/bundler-webpack/compiled/webpack-bundle-analyzer");
function addBundleAnalyzerPlugin(opts) {
    return __awaiter(this, void 0, void 0, function* () {
        const { config } = opts;
        config.plugin('webpack-bundle-analyzer').use(webpack_bundle_analyzer_1.BundleAnalyzerPlugin, [
            // https://github.com/webpack-contrib/webpack-bundle-analyzer
            {
                analyzerMode: 'server',
                analyzerPort: process.env.ANALYZE_PORT || 8888,
                openAnalyzer: false,
                logLevel: 'info',
                defaultSizes: 'parsed',
            },
        ]);
    });
}
exports.addBundleAnalyzerPlugin = addBundleAnalyzerPlugin;
